### PR TITLE
feat(compiler): inflight initilizers and inflight fields

### PR DESCRIPTION
**Pull request by yoav-steinberg**
see #1567 #1596 #67

```wing
bring cloud;

resource Foo {
  inflight var _tmp: num;
  init() { }
  inflight init() {
    this._tmp = 0;
  }
  inflight inc_tmp() {
    this._tmp = this._tmp + 1;
  }
}

let a = new Foo();
new cloud.Function(inflight () => {
  a.inc_tmp();
  log("hey ${a._tmp}");
}) as "test";
```

- [x] missing tests

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: examples/tests/invalid/reassign_to_nonreassignable.w [hunk 1]
```diff
@@ -5,13 +5,23 @@ x = x + 1;
 // Assign to non-reassignable field
 resource R {
   f: num;
+  inflight_f: num;
```
Comment (eladb) [inline examples/tests/invalid/reassign_to_nonreassignable.w:8]: Can you please add the errors as comments? It's very helpful

Comment (eladb) [inline examples/tests/invalid/reassign_to_nonreassignable.w:8]:
```suggestion
  not_inflight_f: num;
```

File: examples/tests/invalid/resource_init.w [hunk 2]
```diff
@@ -0,0 +1,11 @@
+resource R {
+  init() {}
+  init() {}
+//^-- can't have multiple inits
+  inflight init() {}
+  inflight init() {}
+        // ^-- can't have multiple inflight inits
+
+  inflight init(x: num) {}
+             // ^-- can't have inflight init with args
```
Comment (eladb) [inline examples/tests/invalid/resource_init.w:10]: nice

File: libs/wingc/src/jsify.rs [hunk 3]
```diff
@@ -1108,7 +1127,31 @@ impl<'a> JSifier<'a> {
...
+					def,
+					&JSifyContext {
+						in_json: context.in_json.clone(),
+						phase: def.signature.phase,
+					},
+				)
+			})
+			.collect_vec();
+
+		if let Some(inflight_init) = inflight_init {
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:1144]:
```suggestion
                // add the inflight init method to the list of client methods
		if let Some(inflight_init) = inflight_init {
```

File: libs/wingc/src/jsify.rs [hunk 4]
```diff
@@ -1066,23 +1077,31 @@ impl<'a> JSifier<'a> {
...
 			.join(", ");
-		formatdoc!("
+		formatdoc!(
+			"
 			_toInflight() {{
 				{inner_clients}
 				const self_client_path = {client_path};
-				return {STDLIB}.core.NodeJsCode.fromInline(`(new (require(\"${{self_client_path}}\")).{resource_name}({{{captured_fields}}}))`);
+				return {STDLIB}.core.NodeJsCode.fromInline(`(
+					await (async () => {{
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:1086]: Interesting... This implies that `_toInflight()` is called from an async scope. Is that the case? 

Feels like this is an assumption we should model somehow into the API. Otherwise, someone will assume in the future that it's a sync call.

@Chriscbr curious what you think...

File: libs/wingc/src/jsify.rs [hunk 5]
```diff
@@ -1066,23 +1077,31 @@ impl<'a> JSifier<'a> {
...
+		formatdoc!(
+			"
 			_toInflight() {{
 				{inner_clients}
 				const self_client_path = {client_path};
-				return {STDLIB}.core.NodeJsCode.fromInline(`(new (require(\"${{self_client_path}}\")).{resource_name}({{{captured_fields}}}))`);
+				return {STDLIB}.core.NodeJsCode.fromInline(`(
+					await (async () => {{ 
+						const tmp = new (require(\"${{self_client_path}}\")).{resource_name}({{{captured_fields}}}); 
+						if (tmp.{WING_INFLIGHT_INIT_NAME} !== undefined) await tmp.{WING_INFLIGHT_INIT_NAME}();
```
Comment (eladb) [inline libs/wingc/src/jsify.rs:1088]:
```suggestion
						if (tmp.{WING_INFLIGHT_INIT_NAME}) { await tmp.{WING_INFLIGHT_INIT_NAME}(); }
```

File: examples/tests/invalid/reassign_to_nonreassignable.w (hunk 1 above)
Comment (yoav-steinberg) [reply examples/tests/invalid/reassign_to_nonreassignable.w]: fixed

File: libs/wingc/src/type_check.rs [hunk 6]
```diff
@@ -2070,6 +2071,7 @@ impl<'a> TypeChecker<'a> {
 				implements,
 				constructor,
```
Comment (staycoolcall911) [inline libs/wingc/src/type_check.rs:2072]: Consider renaming:
```suggestion
				initializer,
```

**Review by staycoolcall911 — COMMENTED** [review]
Looks great - just one small comment from me.

File: libs/wingc/src/type_check.rs [hunk 7]
```diff
@@ -159,7 +159,8 @@ pub enum Type {
 	Enum(Enum),
 }
 
-const WING_CONSTRUCTOR_NAME: &'static str = "init";
+const WING_INIT_NAME: &'static str = "init";
+pub const WING_INFLIGHT_INIT_NAME: &'static str = "$inflight_init";
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:163]: (optional)
```suggestion
const CLASS_INIT_NAME: &'static str = "init";
pub const CLASS_INFLIGHT_INIT_NAME: &'static str = "$inflight_init";
```

File: libs/wingc/src/type_check.rs [hunk 8]
```diff
@@ -2068,8 +2069,9 @@ impl<'a> TypeChecker<'a> {
 				methods,
 				parent,
 				implements,
-				constructor,
+				initializer: constructor,
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:2072]: style: rename local variable to match (in VS code you can right click on "constructor" in VS Code and click "rename symbol" and type out "initializer" and it should automatically fix it)
```suggestion
				initializer,
```

File: libs/wingc/src/type_check.rs [hunk 9]
```diff
@@ -2172,18 +2174,45 @@ impl<'a> TypeChecker<'a> {
 
 				// Add the constructor to the class env
 				let init_symb = Symbol {
-					name: WING_CONSTRUCTOR_NAME.into(),
-					span: name.span.clone(),
+					name: WING_INIT_NAME.into(),
+					span: name.span.clone(), // TODO: change to init's span
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:2178]: Opening a thread in case this needs to be fixed before merge

File: libs/wingc/src/type_check.rs [hunk 10]
```diff
@@ -2172,18 +2174,45 @@ impl<'a> TypeChecker<'a> {
...
 				};
 				self.add_method_to_class_env(&constructor.signature, env, None, &mut class_env, &init_symb);
 
+				// Add the inflight initializer to the class env
+				let inflight_init_symb = Symbol {
+					name: WING_INFLIGHT_INIT_NAME.into(),
+					span: name.span.clone(), // TODO: change to inflight init's span
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:2185]: Fix

File: libs/wingc/src/visit.rs [hunk 11]
```diff
@@ -206,7 +206,7 @@ where
 {
 	v.visit_symbol(&node.name);
 
-	v.visit_constructor(&node.constructor);
+	v.visit_constructor(&node.initializer);
```
Comment (Chriscbr) [inline libs/wingc/src/visit.rs:209]: We also need to visit the inflight initializer?

(I see this is something we should automate, I've created an issue for it for now https://github.com/winglang/wing/issues/2043)

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 12]
```diff
@@ -838,6 +838,44 @@ exports[`resource_inflight.w > stderr 1`] = `
...
+  |
+6 |   inflight init() {}
+  |   ^^^^^^^^^^^^^^^^^^ Multiple inflight initializers defined in class R
+
+
+error: Unexpected 'ERROR'.
+  --> ../../../examples/tests/invalid/resource_init.w:9:12
+  |
+9 |   inflight init(x: num) {}
+  |            ^^^^^ Unexpected 'ERROR'.
```
Comment (Chriscbr) [inline tools/hangar/__snapshots__/invalid.ts.snap:860]: Can we improve this error message (or create an issue if it seems out of scope of this PR)

File: libs/wingc/src/jsify.rs (hunk 4 above)
Comment (Chriscbr) [reply libs/wingc/src/jsify.rs:1095]: `_toInflight()` is still sync -- do you mean to say the code fragment returned by `toInflight()` is called from an async scope? Yes, I agree it feels like should that should be documented it in the `_toInflight()` API since it means you can't call initialize one of the clients from an ordinary JavaScript constructor for example.

Technically I guess this isn't a big change since that is already how we are invoking some inflight code right now (see below), but something to keep our eyes on if we ever need to initialize inflight code from plain javascript constructors for some reason.

```js
// this is the structure of the code that the SDK bundles into an AWS Lambda function
exports.handler = async function(event) {
  return await (< inflight._toInflight().text >).handle(event);
};
```

File: examples/tests/valid/resource.w [hunk 13]
```diff
@@ -3,11 +3,16 @@ bring cloud;
 // User defined resource
```
Comment (Chriscbr) [inline examples/tests/valid/resource.w:3]: Can we add an example where a user defined resource has an inflight init function, and it has a child resource that also has an inflight init function (and validate that it's initialized properly etc.)? It's important to check that these behaviors work with more than one layer of composition

File: examples/tests/invalid/reassign_to_nonreassignable.w [hunk 14]
```diff
@@ -5,17 +5,30 @@ x = x + 1;
...
   init() {
     this.f = 1;
   }
+
+  inflight init() {
+    this.inflight_f = 1;
+  }
   
   inc() {
     this.f = this.f + 1;
```
Comment (Chriscbr) [inline examples/tests/invalid/reassign_to_nonreassignable.w:19]: Maybe we could also test a case like
`this.child_resource.f = 1 // this.child_resource.f is not reassignable`

**Review by Chriscbr — COMMENTED** [review]
Looks really promising Yoav :)

File: libs/wingc/src/ast.rs [hunk 15]
```diff
@@ -326,7 +337,8 @@ pub struct Class {
 	pub name: Symbol,
 	pub fields: Vec<ClassField>,
 	pub methods: Vec<(Symbol, FunctionDefinition)>,
-	pub constructor: Constructor,
+	pub initializer: Constructor,
+	pub inflight_initializer: Option<FunctionDefinition>,
```
[the remark this answers is not in the document]
Comment (eladb) [reply libs/wingc/src/ast.rs:341]: +1 on always having an inflight initializer (even if it's empty). I agree that if we can reduce the amount of `Option<T>`s in our code, by definition we will have less code branches to handle downstream.

[the remark this answers is not in the document]
Comment (MarkMcCulloh) [reply libs/wingc/src/ast.rs:341]: > Can we avoid the Option indirection?

The Option is technically accurate though, no? I understand that the end result is the same (for now), but wouldn't we rather the model itself be correct? Being forced to handle this case could even be a little micro-optimization, at the cost of an extra line or two of code.

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 16]
```diff
@@ -838,6 +838,44 @@ exports[`resource_inflight.w > stderr 1`] = `
...
+
+
+error: Unexpected 'identifier'.
+  --> ../../../examples/tests/invalid/resource_init.w:9:17
+  |
+9 |   inflight init(x: num) {}
+  |                 ^ Unexpected 'identifier'.
```
[the remark this answers is not in the document]
Comment (yoav-steinberg) [reply tools/hangar/__snapshots__/invalid.ts.snap:867]: Yes, good idea.

File: examples/tests/valid/resource.w (hunk 13 above)
[the remark this answers is not in the document]
Comment (yoav-steinberg) [reply examples/tests/valid/resource.w:3]: it works, i'll add it to the test.

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 12 above)
Comment (yoav-steinberg) [reply tools/hangar/__snapshots__/invalid.ts.snap]: fixed by adding inflight init parameters as valid grammar and error reporting in _parser.rs_

File: examples/tests/valid/resource.w (hunk 13 above)
Comment (yoav-steinberg) [reply examples/tests/valid/resource.w:3]: I agree, but we currently don't support a user defined resource inheriting from another user defined resource because of no `super()` support: #1123
I'll add a comment about that in the relevant issue.

Comment (Chriscbr) [reply examples/tests/valid/resource.w:3]: I think inheritance may be a separate use case than having one resource containing another resource

Comment (yoav-steinberg) [reply examples/tests/valid/resource.w:3]: > I think inheritance may be a separate use case than having one resource containing another resource

I think that case is covered in the _resource.w_ test.

Comment (Chriscbr) [reply examples/tests/valid/resource.w:3]: You're right, I missed it

**Comment by eladb** [comment]
Cool!
